File: app/services/otx.py
import httpx
from app.config import settings
from app.utils.normalize import utc_now
from app.models.shared import SourceMetadata

OTX_URL = "https://otx.alienvault.com/api/v1/indicators/IPv4"

async def fetch_otx(ip: str) -> tuple[dict | None, SourceMetadata]:
    source = SourceMetadata(
        source="alienvault_otx",
        retrieved_at=utc_now(),
        success=False
    )
    try:
        async with httpx.AsyncClient() as client:
            response = await client.get(
                f"{OTX_URL}/{ip}/general",
                headers={
                    "X-OTX-API-KEY": settings.otx_api_key,
                    "Accept": "application/json"
                },
                timeout=5.0
            )
            response.raise_for_status()
            raw = response.json()

            normalized = {
                "pulse_count": raw.get("pulse_info", {}).get("count", 0),
                "malware_families": [
                    p.get("name") for p in
                    raw.get("pulse_info", {}).get("pulses", [])[:5]
                ],
                "country": raw.get("country_name"),
                "asn": raw.get("asn"),
            }

            source.success = True
            return normalized, source

    except httpx.TimeoutException:
        source.error = "Request timed out"
        return None, source
    except httpx.HTTPStatusError as e:
        source.error = f"HTTP {e.response.status_code}"
        return None, source
    except Exception as e:
        source.error = str(e)
        return None, source

File: scanner-go/internal/services/otx.go
package services

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/Ahlyx/scanner-go/internal/models"
)

const (
	otxIPURL     = "https://otx.alienvault.com/api/v1/indicators/IPv4"
	otxDomainURL = "https://otx.alienvault.com/api/v1/indicators/domain"
)

// otxRaw is the subset of OTX response we parse (data isn't surfaced in the
// response models but the SourceMetadata is included in sources[]).
type otxRaw struct {
	PulseInfo struct {
		Count  int `json:"count"`
		Pulses []struct {
			Name string `json:"name"`
		} `json:"pulses"`
	} `json:"pulse_info"`
}

func otxGet(apiKey, url string) ([]byte, error, int) {
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, err, 0
	}
	req.Header.Set("X-OTX-API-KEY", apiKey)
	req.Header.Set("Accept", "application/json")

	// OTX has a 5s timeout in the Python version.
	client := &http.Client{Timeout: 5 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err, 0
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	return body, err, resp.StatusCode
}

// FetchOTXIP fetches AlienVault OTX data for an IP. The returned struct is
// ignored by the IP handler (OTX data isn't in IPResponse fields); only the
// SourceMetadata is used.
func FetchOTXIP(apiKey, ip string) (interface{}, models.SourceMetadata) {
	meta := models.SourceMetadata{Source: "alienvault_otx", RetrievedAt: time.Now().UTC()}

	acquireSem()
	defer releaseSem()

	body, err, status := otxGet(apiKey, fmt.Sprintf("%s/%s/general", otxIPURL, ip))
	if err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}
	if status != http.StatusOK {
		meta.Error = ptr(fmt.Sprintf("HTTP %d", status))
		return nil, meta
	}

	var raw otxRaw
	if err := json.Unmarshal(body, &raw); err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}

	meta.Success = true
	return &raw, meta
}

// FetchOTXDomain fetches AlienVault OTX data for a domain. Similarly, only
// SourceMetadata is used in DomainResponse.
func FetchOTXDomain(apiKey, domain string) (interface{}, models.SourceMetadata) {
	meta := models.SourceMetadata{Source: "alienvault_otx", RetrievedAt: time.Now().UTC()}

	acquireSem()
	defer releaseSem()

	body, err, status := otxGet(apiKey, fmt.Sprintf("%s/%s/general", otxDomainURL, domain))
	if err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}
	if status != http.StatusOK {
		meta.Error = ptr(fmt.Sprintf("HTTP %d", status))
		return nil, meta
	}

	var raw otxRaw
	if err := json.Unmarshal(body, &raw); err != nil {
		meta.Error = ptr(err.Error())
		return nil, meta
	}

	meta.Success = true
	return &raw, meta
}
